File: iosTestApp/App/DesignSystem/ListScreenScaffold.swift
import UIKit

/// A screen whose middle is a list and whose ends are not.
///
/// `ScreenScaffold` scrolls everything together, which is right for a form and
/// wrong for a list: on History it put the selected run's detail *below* the
/// run list, so choosing a run scrolled its own detail off the bottom, and the
/// more history a user had the further away the answer moved. Here the header
/// and the detail stay put and only the rows move.
///
/// Mirror of the Android `ListScreenScaffold`.
final class ListScreenScaffold: UIView {

    private let headerStack = UIStackView()
    private let listStack = UIStackView()
    private let actionStack = UIStackView()
    private let scrollView = UIScrollView()

    init() {
        super.init(frame: .zero)
        backgroundColor = Theme.Color.background

        [headerStack, listStack, actionStack].forEach {
            $0.axis = .vertical
            $0.translatesAutoresizingMaskIntoConstraints = false
        }
        headerStack.spacing = Theme.Space.m
        listStack.spacing = Theme.Space.s
        actionStack.spacing = Theme.Space.s

        headerStack.isLayoutMarginsRelativeArrangement = true
        headerStack.layoutMargins = UIEdgeInsets(
            top: Theme.Space.l, left: Theme.Space.l, bottom: Theme.Space.m, right: Theme.Space.l
        )
        actionStack.isLayoutMarginsRelativeArrangement = true
        actionStack.layoutMargins = UIEdgeInsets(
            top: Theme.Space.m, left: Theme.Space.l, bottom: Theme.Space.m, right: Theme.Space.l
        )

        scrollView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.alwaysBounceVertical = true
        scrollView.addSubview(listStack)

        addSubview(headerStack)
        addSubview(scrollView)
        addSubview(actionStack)

        NSLayoutConstraint.activate([
            headerStack.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor),
            headerStack.leadingAnchor.constraint(equalTo: leadingAnchor),
            headerStack.trailingAnchor.constraint(equalTo: trailingAnchor),

            scrollView.topAnchor.constraint(equalTo: headerStack.bottomAnchor),
            scrollView.leadingAnchor.constraint(equalTo: leadingAnchor),
            scrollView.trailingAnchor.constraint(equalTo: trailingAnchor),
            scrollView.bottomAnchor.constraint(equalTo: actionStack.topAnchor),
            // Without a floor, a tall header plus a long selected-run detail
            // can squeeze the list to nothing on a short screen - and a list
            // of zero height reads as "no history" rather than "scroll me".
            scrollView.heightAnchor.constraint(greaterThanOrEqualToConstant: 140),

            listStack.topAnchor.constraint(equalTo: scrollView.contentLayoutGuide.topAnchor, constant: Theme.Space.s),
            listStack.bottomAnchor.constraint(equalTo: scrollView.contentLayoutGuide.bottomAnchor, constant: -Theme.Space.s),
            listStack.leadingAnchor.constraint(equalTo: scrollView.frameLayoutGuide.leadingAnchor, constant: Theme.Space.l),
            listStack.trailingAnchor.constraint(equalTo: scrollView.frameLayoutGuide.trailingAnchor, constant: -Theme.Space.l),

            actionStack.leadingAnchor.constraint(equalTo: leadingAnchor),
            actionStack.trailingAnchor.constraint(equalTo: trailingAnchor),
            actionStack.bottomAnchor.constraint(equalTo: safeAreaLayoutGuide.bottomAnchor),
        ])
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("created in code") }

    /// Fixed region above the list.
    func addHeader(_ views: UIView...) {
        views.forEach { headerStack.addArrangedSubview($0) }
    }

    /// Replaces the scrolling region's contents.
    func setListItems(_ views: [UIView]) {
        listStack.arrangedSubviews.forEach { $0.removeFromSuperview() }
        views.forEach { listStack.addArrangedSubview($0) }
        scrollView.setContentOffset(.zero, animated: false)
    }

    /// Pinned below the list. Primary action first.
    func addActions(_ views: UIView...) {
        views.forEach { actionStack.addArrangedSubview($0) }
    }
}

File: androidTestApp/src/main/java/edu/gatech/cc/cellwatch/androidtestapp/designsystem/ListScreenScaffold.kt
package edu.gatech.cc.cellwatch.androidtestapp.designsystem

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ScrollView
import edu.gatech.cc.cellwatch.androidtestapp.designsystem.Theme.dp

/**
 * A screen whose middle is a list and whose ends are not.
 *
 * [ScreenScaffold] scrolls everything together, which is right for a form and
 * wrong for a list: on History it put the selected run's detail *below* the
 * run list, so choosing a run scrolled its own detail off the bottom, and the
 * more history a user had the further away the answer moved. Here the header
 * and the detail stay put and only the rows move.
 *
 * Mirror of the iOS `ListScreenScaffold`.
 */
class ListScreenScaffold(context: Context) : LinearLayout(context) {

    private val headerColumn = LinearLayout(context).apply {
        orientation = VERTICAL
        setPadding(
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.M),
        )
    }

    private val listColumn = LinearLayout(context).apply {
        orientation = VERTICAL
        setPadding(
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.S),
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.S),
        )
    }

    private val actionColumn = LinearLayout(context).apply {
        orientation = VERTICAL
        setPadding(
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.M),
            context.dp(Theme.Space.L),
            context.dp(Theme.Space.M),
        )
    }

    private val listScroll = ScrollView(context).apply {
        isFillViewport = true
        // Without a floor, a tall header plus a long selected-run detail can
        // squeeze the list to nothing on a short screen - and a list of zero
        // height reads as "no history" rather than "scroll me".
        minimumHeight = context.dp(140)
        addView(
            listColumn,
            LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT),
        )
    }

    init {
        orientation = VERTICAL
        setBackgroundColor(Theme.Palette.BACKGROUND)

        addView(
            headerColumn,
            LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT),
        )
        addView(listScroll, LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f))
        addView(
            actionColumn,
            LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT),
        )
    }

    /** Fixed region above the list. */
    fun addHeader(vararg views: View) {
        views.forEach { headerColumn.append(it, Theme.Space.M) }
    }

    /** Replaces the scrolling region's contents. */
    fun setListItems(views: List<View>) {
        listColumn.removeAllViews()
        views.forEach { listColumn.append(it, Theme.Space.S) }
        listScroll.scrollTo(0, 0)
    }

    /** Pinned below the list. Primary action first. */
    fun addActions(vararg views: View) {
        views.forEach { actionColumn.append(it, Theme.Space.S) }
    }

    /** Preserves whatever size the component asked for; see [ScreenScaffold]. */
    private fun LinearLayout.append(view: View, spacingDp: Int) {
        val params = view.layoutParams as? LayoutParams
            ?: LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            )
        if (childCount > 0) params.topMargin = context.dp(spacingDp)
        addView(view, params)
    }
}
